### PR TITLE
dsmr support esp-idf

### LINKED ISSUE #2517: DSMR: no gas usage for DSMR2
### The problem

Earlier the DSMR component was updated for usage with the DSMR2 standard. These changes are merged in the dev branch. I tried updating my ESP32 P1 reader with the latest ESPhome dev and now it reports the correct data - except the gas usage which is stays "unknown".

Closed DSMR2 issue: https://github.com/esphome/issues/issues/2363
PR that did not fix this problem: https://github.com/esphome/esphome/pull/2157

### Which version of ESPHome has the issue?

2021.10.0-dev

### What type of installation are you using?

Docker

### Which version of Home Assistant has the issue?

Not relevant

### What platform are you using?

ESP32

### Board

ESP32 wrover

### Component causing the issue

dsmr

### Example YAML snippet

```yaml
#######################################################################
# Use ESP32 Serial2
uart:
  tx_pin: 17
  rx_pin: 16
  baud_rate: 9600
  data_bits: 7
  parity: NONE
  stop_bits: 1

# DSMR2 has no CRC
dsmr:
  crc_check: false

sensor:
  - platform: dsmr
    gas_delivered:
      name: "Gas Delivery Total"
      id: gas_delivery_total
      state_class: "total_increasing"
      unit_of_measurement: "m³"
```

### Anything in the logs that might be useful for us?

```txt
No error messages, just no update for this specific sensor. Same for DSMR version sensor.

Text that DSMR2 produces:
0-2:24.3.0(180822110000)(08)(60)(1)(0-2:24.2.0)(m3)
(04700.692)

Note that the actual sensor data is on the next line.
```

### Additional information

_No response_

### LINKED ISSUE #2638: DSMR:  Gas data does not show up
### The problem

I'm using an Wemos D1 mini and all of the other data shows up, but only the gas data does not.
I've tried playing around with the mbus_id, but this doesn't seem to work for me.
Hooking up an FTDI to a PC and opening the serial port shows this:
`1-3:0.2.8(42)
0-0:1.0.0(211025122251S)
0-0:96.1.1(4530303330303033313730343838313135)
1-0:1.8.1(008537.641*kWh)
1-0:1.8.2(007069.927*kWh)
1-0:2.8.1(001044.662*kWh)
1-0:2.8.2(002515.444*kWh)
0-0:96.14.0(0002)
1-0:1.7.0(00.370*kW)
1-0:2.7.0(00.000*kW)
0-0:96.7.21(00013)
0-0:96.7.9(00009)
1-0:99.97.0(9)(0-0:96.7.19)(210324194538W)(0000004159*s)(190603121040S)(0000000280*s)(190517110331S)(0000000230*s)(181205104105W)(0000002129*s)(160726165207S)(0000000470*s)(160621150145S)(0000007990*s)(160525073504S)(0000000372*s)(160421143231S)(0000000548*s)(160420165926S)(0000000604*s)
1-0:32.32.0(00000)
1-0:52.32.0(00002)
1-0:72.32.0(00004)
1-0:32.36.0(00000)
1-0:52.36.0(00000)
1-0:72.36.0(00000)
0-0:96.13.1()
0-0:96.13.0()
1-0:31.7.0(002*A)
1-0:51.7.0(000*A)
1-0:71.7.0(001*A)
1-0:21.7.0(00.140*kW)
1-0:41.7.0(00.003*kW)
1-0:61.7.0(00.225*kW)
1-0:22.7.0(00.000*kW)
1-0:42.7.0(00.000*kW)
1-0:62.7.0(00.000*kW)
0-2:24.1.0(003)
0-2:96.1.0(4730303332353631323636373632343136)
0-2:24.2.1(211025120000S)(04034.144*m3)
!1EE3
`
From this I can see that the mbus should be set to 2 (which I did), but still nothing.

### Which version of ESPHome has the issue?

2021.10.2

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

core-2021.10.6

### What platform are you using?

ESP8266

### Board

Wemos D1 mini

### Component causing the issue

dsmr

### Example YAML snippet

```yaml
# Example configuration entry
dsmr:
  gas_mbus_id: 2
```

### Anything in the logs that might be useful for us?

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by glmnet**
# What does this implement/fix? 

Removes arduino (and all lib) dependencies and adds support for esp-idf (ESP-C3)

Also
fixes https://github.com/esphome/issues/issues/2638
fixes https://github.com/esphome/issues/issues/2517

Quick description and explanation of changes

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [x] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/dsmr/dsmr.cpp [hunk 1]
```diff
@@ -144,9 +143,10 @@ void Dsmr::receive_encrypted_() {
                          buffer_length - 17);
       delete gcmaes128;  // NOLINT(cppcoreguidelines-owning-memory)
 
-      telegram_len_ = strnlen(this->telegram_, sizeof(this->telegram_));
+      telegram_len_ = packet_size - 30;
       ESP_LOGV(TAG, "Decrypted telegram size: %d bytes", telegram_len_);
-      ESP_LOGVV(TAG, "Decrypted telegram: %s", this->telegram_);
+
+      parse_telegram();
```
Comment (mmakaay) [inline esphome/components/dsmr/dsmr.cpp:149]: This one got duplicated in merging the encryption hotfix.

**Review by mmakaay — COMMENTED** [review]
In general, I'm wondering if it would be possible to place all the newly added files into a `parser/` subdirectory. When looking at the dsmr component folder, the number of files is quite overwhelming now and the ESPHome glue code is intermixed with the plain DSMR parsing functionality. That might make it hard for new coders to find their way in there.

Another thing that pops to mind is that the crypto code might be generic enough to move things to one or more separate modules (like I was requested to do with the MD5 code that implemented for the OTA password and checksum support).
Within the current use case however, that wouldn't add that much, since AFAICS there currently are no other components that would use it. So I would be fine by keeping these libraries contained within the dsmr component folder.

**Update:**
I tested the new code using ESP8266 with my unencrypted DSMR smart meter, and things work as expected 👍 

**Update 2:**
I test the code using ESP8266 with encrypted DSMR data, and also thumbs up.

**Comment by jesserockz** [comment]
@mmakaay Extra subfolders dont work in ESPHome as the code generator only copies the files in the directory and doesnt look at subfolders

**Comment by oxan** [comment]
I'm not too fond of merging all this crypto code into ESPHome itself...

**Comment by OttoWinter** [comment]
> I'm not too fond of merging all this crypto code into ESPHome itself...

I agree.

It looks like this code mostly just copies the code from the dsmr lib into the esphome tree. Why don't we just adjust the dsmr lib to be compatible with esp-idf and use that

In particular we'd just need to make platformio think the crypto lib is compatible with all frameworks ([here](https://github.com/rweather/arduinolibs/blob/master/libraries/Crypto/library.json#L20)); if code changes are needed, I imagine it would be similar as in this PR, no? Maybe reach out to the maintainer or if non-arduino is not a target for them (which would be understandable) fork it.

**Pull request closed without merging**